src/relay/pass/type_solver.cc

```diff
@@ -0,0 +1,166 @@
+/*!
+ *  Copyright (c) 2018 by Contributors
+ * \file type_solver.cc
+ * \brief Type solver implementations.
+ */
+#include <string>
+#include "type_solver.h"
+
+namespace tvm {
+namespace relay {
+
+class TypeSolver::Reporter : public TypeReporterNode {
+ public:
+  explicit Reporter(TypeSolver* solver)
+      : solver_(solver) {}
+
+  void Assign(const Type& dst, const Type& src) final {
+    solver_->Unify(dst, src);
+  }
+
+  void AssertEQ(const ShapeExpr& lhs, const ShapeExpr& rhs) final {
+    // TODO(tqchen)
+  }
+
+ private:
+  TypeSolver* solver_;
+};
+
+// constructor
+TypeSolver::TypeSolver()
+    : reporter_(make_node<Reporter>(this)) {
+}
+
+// destructor
+TypeSolver::~TypeSolver() {
+  // call destructor of all non-POD arena object
+  for (TypeNode* ptr : type_nodes_) {
+    ptr->~TypeNode();
+  }
+  for (RelationNode* ptr : rel_nodes_) {
+    ptr->~RelationNode();
+  }
+}
+
+// Add equality constraint
+Type TypeSolver::Unify(const Type& dst, const Type& src) {
+  // Known limitation
+  // - handle composite types whose component can be unknown.
+  // - handle shape pattern matching
+  TypeNode* lhs = GetTypeNode(dst);
+  TypeNode* rhs = GetTypeNode(src);
+  if (lhs->resolved_type.as<IncompleteTypeNode>()) {
+    MergeFromTo(lhs, rhs);
+    return rhs->resolved_type;
+  } else if (rhs->resolved_type.as<IncompleteTypeNode>()) {
+    MergeFromTo(rhs, lhs);
+    return lhs->resolved_type;
+  } else {
+    lhs->parent = rhs;
```

Comment:
what if it is not nullptr? maybe call find and lhs->parent->parent = rhs?

Reply:
We already call FindRoot in the beginning of the logic so lhs and rhs's parent is always nullptr

Reply:
can we put an assert (or a comment stating the invariant? it is not obvious.)